test/common/upstream/health_check_fuzz.cc

```diff
@@ -0,0 +1,209 @@
+  const int index = (second_host_ && second_host) ? 1 : 0;
+  ENVOY_LOG_MISC(trace, "Created a new stream on host {}", index);
+  expectStreamCreate(index);
+  test_sessions_[index]->interval_timer_->invokeCallback();
+}
+
+void HealthCheckFuzz::clientCreate() { TcpHealthCheckerImplTestBase::expectClientCreate(); }
+
+void HealthCheckFuzz::raiseEvent(test::common::upstream::RaiseEvent event, bool second_host) {
+  Network::ConnectionEvent eventType;
+  switch (event.event_selector_case()) {
+  case test::common::upstream::RaiseEvent::kConnected: {
+    eventType = Network::ConnectionEvent::Connected;
+    break;
+  }
+  case test::common::upstream::RaiseEvent::kRemoteClose: {
+    eventType = Network::ConnectionEvent::RemoteClose;
+    break;
+  }
+  case test::common::upstream::RaiseEvent::kLocalClose: {
+    eventType = Network::ConnectionEvent::LocalClose;
+    break;
+  }
+  default:
+    break;
+  }
+
+  const int index = (second_host_ && second_host) ? 1 : 0;
+  switch (type_) {
+  case HealthCheckFuzz::Type::HTTP: {
+    test_sessions_[index]->client_connection_->raiseEvent(eventType);
+  }
+  case HealthCheckFuzz::Type::TCP: {
+    connection_->raiseEvent(eventType);
+  }
+  default:
+    break;
+  }
+}
+
+void HealthCheckFuzz::replay(test::common::upstream::HealthCheckTestCase input) {
+  for (int i = 0; i < input.actions().size(); ++i) {
+    const auto& event = input.actions(i);
+    ENVOY_LOG_MISC(trace, "Action: {}", event.DebugString());
+    switch (event.action_selector_case()) {
+    case test::common::upstream::Action::kRespond: {
+      switch (type_) {
+      case HealthCheckFuzz::Type::HTTP: {
+        respondHttp(event.respond().http_respond().headers(),
+                    event.respond().http_respond().status(), event.respond().second_host());
+        break;
+      }
+      case HealthCheckFuzz::Type::TCP: {
+        respondTcp(event.respond().tcp_respond().data());
+        break;
+      }
+      default:
+        break;
+      }
+      break;
+    }
+    case test::common::upstream::Action::kStreamCreate: {
+      switch (type_) {
+      case HealthCheckFuzz::Type::HTTP: {
+        streamCreate(event.stream_create().second_host());
+        break;
+      }
+      case HealthCheckFuzz::Type::TCP: {
+        clientCreate();
+        break;
+      }
+      default:
+        break;
+      }
+      break;
+    }
+    case test::common::upstream::Action::kRaiseEvent: {
+      raiseEvent(event.raise_event(), event.raise_event().second_host());
+      break;
+    }
+    case test::common::upstream::Action::kAdvanceTime: {
+      switch (type_) {
+      case HealthCheckFuzz::Type::HTTP: {
+        time_system_.advanceTimeAsync(
+            std::chrono::milliseconds(event.advance_time().ms_advanced()));
+        HttpHealthCheckerImplTestBase::dispatcher_.run(Event::Dispatcher::RunType::NonBlock);
+        break;
+      }
+      case HealthCheckFuzz::Type::TCP: {
+        // TODO: No time system?
```

Comment:
I think there should be a timer involved in TCP (and in any health checker). See [here](https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/health_checker_impl.cc#L537).